File: core/src/net/artux/pda/map/engine/systems/PlayerSystem.java
package net.artux.pda.map.engine.systems;

import com.badlogic.ashley.core.ComponentMapper;
import com.badlogic.ashley.core.Engine;
import com.badlogic.ashley.core.Entity;
import com.badlogic.ashley.core.Family;
import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.Preferences;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.utils.Disposable;

import net.artux.pda.map.di.scope.PerGameMap;
import net.artux.pda.map.engine.components.HealthComponent;
import net.artux.pda.map.engine.components.MoodComponent;
import net.artux.pda.map.engine.components.PositionComponent;
import net.artux.pda.map.engine.components.QuestComponent;
import net.artux.pda.map.engine.components.SpriteComponent;
import net.artux.pda.map.engine.components.VelocityComponent;
import net.artux.pda.map.engine.components.player.PlayerComponent;
import net.artux.pda.model.items.MedicineModel;

import javax.inject.Inject;

@PerGameMap
public class PlayerSystem extends BaseSystem implements Disposable {

    private ComponentMapper<PositionComponent> pm = ComponentMapper.getFor(PositionComponent.class);
    private ComponentMapper<MoodComponent> mm = ComponentMapper.getFor(MoodComponent.class);
    private ComponentMapper<SpriteComponent> sm = ComponentMapper.getFor(SpriteComponent.class);
    private ComponentMapper<HealthComponent> hm = ComponentMapper.getFor(HealthComponent.class);
    private ComponentMapper<PlayerComponent> pmm = ComponentMapper.getFor(PlayerComponent.class);
    private ComponentMapper<VelocityComponent> vcm = ComponentMapper.getFor(VelocityComponent.class);

    private final float pixelsPerMeter = 3f;
    private InteractionSystem interactionSystem;


    @Inject
    public PlayerSystem(InteractionSystem interactionSystem) {
        super(Family.all(PositionComponent.class, QuestComponent.class).get());
        this.interactionSystem = interactionSystem;
    }

    @Override
    public void removedFromEngine(Engine engine) {
        super.removedFromEngine(engine);

        //todo rthis
        interactionSystem.getUserInterface().getGameZone().setVisible(false);
    }

    @Override
    public void addedToEngine(Engine engine) {
        super.addedToEngine(engine);
        loadPreferences();
    }

    public Entity getPlayer() {
        return super.getPlayer();
    }

    @Override
    public void update(float deltaTime) {
        super.update(deltaTime);
        if (getPlayer() != null) {
            MoodComponent moodComponent = mm.get(getPlayer());
            PositionComponent playerPosition = pm.get(getPlayer());
            SpriteComponent spriteComponent = sm.get(getPlayer());
            Entity enemy = moodComponent.enemy;
            Vector2 direction;
            if (enemy == null) {
                direction = vcm.get(getPlayer()).getVelocity();
            } else {
                PositionComponent enemyPosition = pm.get(enemy);
                direction = enemyPosition.getPosition().cpy().sub(playerPosition.getPosition());
            }

            float degrees = (float) (Math.atan2(
                    -direction.x,
                    direction.y
            ) * 180.0d / Math.PI);

            float currentRotation = spriteComponent.getRotation() - 90;
            float alternativeRotation;
            if (currentRotation > 0)
                alternativeRotation = currentRotation - 360;
            else
                alternativeRotation = currentRotation + 360;

            float difference = currentRotation - degrees;
            if (Math.abs(alternativeRotation - degrees) < Math.abs(difference))
                difference = alternativeRotation - degrees;

            if (direction.x != 0 && direction.y != 0) {
                float step = difference * deltaTime * 20;
                spriteComponent.setRotation(spriteComponent.getRotation() - step);
            }

        }
    }

    @Override
    protected void processEntity(Entity entity, float deltaTime) {

    }

    public PlayerComponent getPlayerComponent() {
        return pmm.get(getPlayer());
    }

    public HealthComponent getHealthComponent() {
        return hm.get(getPlayer());
    }

    public Vector2 getPosition() {
        PositionComponent playerPosition = pm.get(getPlayer());
        return playerPosition.getPosition();
    }

    public int getDistance() {
        if (getEntities() != null && getEntities().size() > 0) {
            return (int) (getPosition().dst(pm.get(getEntities().first()).getPosition()) / pixelsPerMeter);//TODO select target
        } else return -1;
    }

    @Override
    public void dispose() {
        savePreferences();
    }

    public void savePreferences() {
        if (getPlayer() != null) {
            Preferences preferences = Gdx.app.getPreferences("player");

            HealthComponent healthComponent = getHealthComponent();
            preferences.putFloat("health", healthComponent.value);
            preferences.putFloat("radiation", healthComponent.radiation);

            preferences.flush();
        }
    }

    private void loadPreferences() {
        if (getPlayer() != null) {
            Preferences preferences = Gdx.app.getPreferences("player");

            HealthComponent healthComponent = getHealthComponent();
            healthComponent.value = preferences.getFloat("health", 100);
            if (healthComponent.isDead())
                healthComponent.value = 50;
            healthComponent.radiation = preferences.getFloat("radiation", 0);
        }
    }


    public interface MedicineListener {
        void treat(MedicineModel model);
    }
}

File: core/src/net/artux/pda/map/engine/ecs/systems/player/PlayerSystem.kt
package net.artux.pda.map.engine.ecs.systems.player

import com.badlogic.ashley.core.*
import com.badlogic.gdx.Gdx
import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.utils.Disposable
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import net.artux.pda.map.repository.DataRepository
import net.artux.pda.map.engine.ecs.components.BodyComponent
import net.artux.pda.map.engine.ecs.components.HealthComponent
import net.artux.pda.map.engine.ecs.components.MoodComponent
import net.artux.pda.map.engine.ecs.components.SpriteComponent
import net.artux.pda.map.engine.ecs.components.player.PlayerComponent
import net.artux.pda.map.engine.ecs.systems.BaseSystem
import net.artux.pda.map.utils.di.scope.PerGameMap
import net.artux.pda.map.view.UserInterface
import net.artux.pda.model.quest.story.StoryDataModel
import javax.inject.Inject
import kotlin.math.abs
import kotlin.math.atan2

@PerGameMap
class PlayerSystem @Inject constructor(
    val dataRepository: DataRepository,
    val userInterface: UserInterface,
    val playerMovingSystem: PlayerMovingSystem
) :
    BaseSystem(Family.one().get()), Disposable {
    private val pm = ComponentMapper.getFor(
        BodyComponent::class.java
    )
    private val mm = ComponentMapper.getFor(
        MoodComponent::class.java
    )
    private val sm = ComponentMapper.getFor(
        SpriteComponent::class.java
    )
    private val hm = ComponentMapper.getFor(
        HealthComponent::class.java
    )
    lateinit var lastDataModel: StoryDataModel

    init {
        CoroutineScope(Dispatchers.Main).launch {
            dataRepository.storyDataModelFlow.collect {
                lastDataModel = it
            }
        }
    }

    override fun removedFromEngine(engine: Engine) {
        super.removedFromEngine(engine)
        userInterface.gameZone.isVisible = false
    }

    override fun addedToEngine(engine: Engine) {
        super.addedToEngine(engine)

        engine.addEntityListener(
            Family.one(PlayerComponent::class.java).get(),
            object : EntityListener {
                override fun entityAdded(entity: Entity?) {
                    loadPreferences()
                }

                override fun entityRemoved(entity: Entity?) {

                }

            })
    }

    public override fun getPlayer(): Entity? {
        return super.getPlayer()
    }

    val direction: Vector2 = Vector2()

    override fun update(deltaTime: Float) {
        super.update(deltaTime)
        if (isPlayerActive) {
            val moodComponent = mm[player]
            val playerPosition = pm[player]
            val spriteComponent = sm[player]
            val enemy = moodComponent.enemy
            if (enemy == null) {
                direction.set(playerMovingSystem.velocity)
            } else {
                val enemyPosition = pm[enemy]
                direction.set(
                    enemyPosition.position.x - playerPosition.position.x,
                    enemyPosition.position.y - playerPosition.position.y
                )
            }

            val degrees = (atan2(
                -direction.x.toDouble(),
                direction.y
                    .toDouble()
            ) * 180.0 / Math.PI).toFloat()

            val currentRotation = spriteComponent.rotation - 90
            val alternativeRotation =
                if (currentRotation > 0) currentRotation - 360 else currentRotation + 360
            var difference = currentRotation - degrees
            if (abs(alternativeRotation - degrees) < abs(difference)) difference =
                alternativeRotation - degrees
            if (direction.x != 0f && direction.y != 0f) {
                val step = difference * deltaTime * 20
                spriteComponent.rotation = spriteComponent.rotation - step
            }
        }
    }

    override fun processEntity(entity: Entity, deltaTime: Float) {}
    val healthComponent: HealthComponent get() = hm[player]
    val position: Vector2
        get() = pm[player].position

    override fun dispose() {
        if(isPlayerActive)
            savePreferences()
        else{
            val preferences = Gdx.app.getPreferences("player")
            preferences.clear()
            preferences.flush()
        }
    }

    private fun savePreferences() {
        if (isPlayerActive) {
            val preferences = Gdx.app.getPreferences("player")
            val healthComponent = healthComponent
            preferences.putFloat("health", healthComponent.health)
            preferences.putFloat("radiation", healthComponent.radiation)
            preferences.flush()
        }
    }

    private fun loadPreferences() {
        if (isPlayerActive) {
            val preferences = Gdx.app.getPreferences("player")
            val healthComponent = healthComponent
            healthComponent.health = preferences.getFloat("health", 100f)
            if (healthComponent.isDead()) healthComponent.health = 50f
            healthComponent.radiation = preferences.getFloat("radiation", 0f)
        }
    }
}